### PR TITLE
AirfRANS: softer gc=0.5 on EMA champion (gc reduction transfer from TF)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

TandemFoil experiments showed a clear monotonic improvement as grad-clip was softened: gc=1.0 → gc=0.5+EMA improved by -13.8%, then gc=0.3+EMA improved a further -2.8%. The AirfRANS EMA champion (PR #3050, stark) currently uses gc=1.0. We hypothesise that the same gc-reduction transfer applies to AirfRANS: reducing gc from 1.0 → 0.5 will allow the optimizer to take larger gradient steps in the flat regions of the AirfRANS loss landscape, improving surface and volume MSE.

The TandemFoil gc pattern is strong enough evidence to motivate a direct cross-dataset transfer test. If gc=0.5 beats gc=1.0 on AirfRANS, we follow up with gc=0.3.

## Instructions

Start from the AirfRANS EMA champion config (PR #3050) and change only `--grad-clip` from 1.0 to 0.5. Everything else stays identical.

```bash
cd target/icml2026
python train.py \
  --dataset airfrans \
  --airfrans-task full \
  --optimizer adamw \
  --lr 6e-4 \
  --cosine-t-max 50 \
  --grad-clip 0.5 \
  --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 2 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --epochs 999 \
  --ema-decay 0.999 \
  --wandb-group af-gc-sweep
```

Key change vs champion: `--grad-clip 0.5` (was 1.0). All other flags are identical to PR #3050.

**What to report:** Compare `val_primary/surface_mse` and `full_val/volume_mse` against the gc=1.0 champion below. Report best checkpoint metrics (not terminal epoch).

## Baseline

Current AirfRANS champion — PR #3050 (stark/af-ema-champion):
- `val_primary/surface_mse`: **0.000459** (ep771)
- `full_val/volume_mse`: **0.002777**
- Config: 2L/256d/4H, AdamW lr=6e-4, cosine T_max=50, **gc=1.0**, WD=1e-2, EMA=0.999, Fourier enabled
- W&B run: `z6pry4b9` — https://wandb.ai/wandb/senpai/runs/z6pry4b9
- Reproduce champion: `cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --ema-decay 0.999`

**Target to beat:** surface_mse < 0.000459 and/or vol_mse < 0.002777

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT edward:

## Results

### gc=0.5 + EMA=0.999 (AirfRANS)

| Metric | gc=0.5 (ep356) | gc=1.0 champion (ep771) | Delta |
|--------|----------------|------------------------|-------|
| `val_primary/surface_mse` | **0.000451** | 0.000459 | **-1.8%** ✅ |
| `full_val/volume_mse` | 0.00585 | **0.002777** | **+110.7%** ❌ |
| `test_primary/surface_mse` | 0.000938 | — | — |
| `full_test/volume_mse` | 0.00696 | — | — |

**W&B run:** `7zrs1bxp` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/7zrs1bxp
**Peak memory:** ~46 GB (GPU 0)
**Epochs:** 763 total, best at epoch 356 (360-min budget)
**Grad clip events:** 345/360 batches clipped per epoch, mean grad norm 12.6

**Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 --grad-clip 0.5 --weight-decay 1e-2 --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 \
  --ema-decay 0.999 --wandb-group af-gc-sweep \
  --agent edward --wandb-name "edward/af-gc05-ema"
```

### What happened

**Mixed result — marginal surface win, significant volume regression.**

gc=0.5 achieved a small surface MSE improvement (-1.8%, 0.000451 vs 0.000459), but volume MSE more than doubled (+110.7%, 0.00585 vs 0.002777). The gc-reduction transfer from TandemFoil does **not** cleanly apply to AirfRANS.

Key observation: the best checkpoint was at epoch 356 out of 763 total — meaning the model regressed badly for the second half of training. The gc=1.0 champion found its best at ep771 with no such regression. This suggests gc=0.5 introduces optimization instability on AirfRANS that the longer cosine cycle (T_max=50) cannot fully compensate for. The volume channels (Ux, Uy, nut) appear particularly sensitive to the looser gradient clipping.

The TandemFoil→AirfRANS transfer hypothesis is **falsified** for gc reduction: the gc=0.5 instability effect dominates any basin-deepening benefit on this dataset.

### Bug fix

Fixed a `primary_metric_key` UnboundLocalError in train.py — a local variable shadowed the function of the same name (lines 1640-1643 removed, line 1716 updated to use `best_val_primary_metric_name`). This bug would have blocked all training runs.

### Suggested follow-ups

- **gc=0.3 is unlikely to help** — if gc=0.5 already destabilizes AirfRANS, gc=0.3 will be worse.
- The volume regression suggests AirfRANS benefits from tighter gradient control than TandemFoil. The datasets have fundamentally different optimization landscapes (4-channel volume prediction vs surface-only pressure).
- A potential middle ground: gc=0.75 (between champion 1.0 and tested 0.5) might capture a small surface benefit without the volume blowup.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3156 because ADVISOR: Closing — gc=0.5 surface=0.000451, vol=0.005850 (2.9x worse). Softer gradient clipping severely degrades volume fidelity. Champion gc=1.0 is correct for AF. Edward reassigned..
